tests/test_schema.py

```diff
@@ -2265,3 +2265,68 @@ class ModelModified(BaseModel):
         },
         'required': ['data', 'data1'],
     }
+
+
+@pytest.mark.skipif(not Literal, reason='typing_extensions not installed and python version < 3.8')
+def test_discriminated_union():
+    class Cat(BaseModel):
+        pet_type: Literal['cat']
+
+    class Dog(BaseModel):
+        pet_type: Literal['dog']
+
+    class Lizard(BaseModel):
+        pet_type: Literal['reptile', 'lizard']
+
+    class Model(BaseModel):
+        pet: Union[Cat, Dog, Lizard] = Field(..., discriminator='pet_type')
+
+    assert Model.schema() == {
+        'title': 'Model',
+        'type': 'object',
+        'properties': {
+            'pet': {
+                'title': 'Pet',
+                'anyOf': [
+                    {'$ref': '#/definitions/Cat'},
+                    {'$ref': '#/definitions/Dog'},
+                    {'$ref': '#/definitions/Lizard'},
+                ],
+                'discriminator': {
+                    'propertyName': 'pet_type',
+                    'mapping': {
+                        'cat': '#/definitions/Cat',
+                        'dog': '#/definitions/Dog',
+                        'lizard': '#/definitions/Lizard',
+                        'reptile': '#/definitions/Lizard',
+                    },
+                },
```

Comment:
`discriminator` is supported from OpenAPI 3.x on but that one also seems to use a different way of storing definitions. Seems that they were moved from `definitions` to `components/schema`.
Also in the end Pydantic seems to target JSON schema? Just wondering how this all fits 😅

Reply:
My main concern was to speed up the validation and have a nice and clean error when not valid.
The JSON schema support is messy and I tried to implement the basic one for FastAPI more than myself.
Currently this is the result

| discriminator  | code  | ReDoc  |
|---|---|---|
| without  | ![without_discriminator_code](https://user-images.githubusercontent.com/18406791/108592379-e7a65800-736d-11eb-84bb-e508f04c654e.png)  | ![without_discriminator_redoc](https://user-images.githubusercontent.com/18406791/108592380-e83eee80-736d-11eb-9486-d79943084683.png)
  |
| with  | ![with_discriminator_code](https://user-images.githubusercontent.com/18406791/108592377-e6752b00-736d-11eb-9ce6-82b833d11382.png)  | ![with_discriminator_redoc](https://user-images.githubusercontent.com/18406791/108592378-e70dc180-736d-11eb-8eb3-22caab9afd68.png)  |

I'm not sure I want to put more effort on it without clear expectations.
If the current behaviour is wrong and someone can produce a full example of FastAPI  code with expected ReDoc, I will try to make it work 👍